src/desktop/space/mod.rs

```diff
@@ -358,826 +316,263 @@ impl Space {
+    OutputRenderElements<'a, R, E, C>;
+    Space=SpaceRenderElements<R, E>,
+    Custom=&'a C,
+}
 
-        fn type_of(&self) -> std::any::TypeId {
-            match self {
-                $(
-                    $(
-                        #[$meta]
-                    )*
-                    Self::$body(x) => $crate::custom_elements_internal!(@call $renderer $(as $other_renderer)?; type_of; x)
-                ),*,
-                Self::_GenericCatcher(_) => unreachable!(),
-            }
+/// Get the render elements for a specific output
+pub fn space_render_elements<
+    'a,
+    #[cfg(feature = "wayland_frontend")] R: Renderer + ImportAll,
+    #[cfg(not(feature = "wayland_frontend"))] R: Renderer,
+    E: SpaceElement + PartialEq + AsRenderElements<R>,
+>(
+    spaces: &[&'a Space<E>],
+    output: &Output,
+) -> Result<Vec<SpaceRenderElements<R, <E as AsRenderElements<R>>::RenderElement>>, OutputNoMode>
+where
+    <R as Renderer>::TextureId: Texture + 'static,
+    <E as AsRenderElements<R>>::RenderElement: 'a,
+    SpaceRenderElements<R, <E as AsRenderElements<R>>::RenderElement>:
+        From<Wrap<<E as AsRenderElements<R>>::RenderElement>>,
+{
+    let mut render_elements = Vec::new();
+
+    for space in spaces {
+        match space.elements_for_output(output) {
+            Ok(elements) => render_elements.extend(elements),
+            Err(OutputError::Unmapped) => {}
+            Err(OutputError::NoMode(_)) => return Err(OutputNoMode),
         }
+    }
 
-        fn location(&self, scale: impl Into<$crate::utils::Scale<f64>>) -> $crate::utils::Point<f64, $crate::utils::Physical> {
-            match self {
-                $(
-                    $(
-                        #[$meta]
-                    )*
-                    Self::$body(x) => $crate::custom_elements_internal!(@call $renderer $(as $other_renderer)?; location; x, scale)
-                ),*,
-                Self::_GenericCatcher(_) => unreachable!(),
-            }
-        }
+    Ok(render_elements)
+}
 
-        fn geometry(&self, scale: impl Into<$crate::utils::Scale<f64>>) -> $crate::utils::Rectangle<i32, $crate::utils::Physical> {
-            match self {
-                $(
-                    $(
-                        #[$meta]
-                    )*
-                    Self::$body(x) => $crate::custom_elements_internal!(@call $renderer $(as $other_renderer)?; geometry; x, scale)
-                ),*,
-                Self::_GenericCatcher(_) => unreachable!(),
-            }
-        }
+/// Render a output
+#[allow(clippy::too_many_arguments)]
+pub fn render_output<
+    'a,
+    #[cfg(feature = "wayland_frontend")] R: Renderer + ImportAll,
+    #[cfg(not(feature = "wayland_frontend"))] R: Renderer,
+    C: RenderElement<R>,
+    E: SpaceElement + PartialEq + AsRenderElements<R>,
+    L: Into<Option<slog::Logger>>,
+>(
+    output: &Output,
+    renderer: &mut R,
+    age: usize,
+    spaces: &[&'a Space<E>],
```

Comment:
&[&Space] seems difficult to use from API perspective. From my perspective, I'd like &[Space], but even [&Space] seems a bit more idiomatic.

Reply:
`&[Space]` would mean you have to have ownership over the spaces inside the collection.
If you don't happen to store them in a Vec<Space> anyway, this is cumbersome for the users.

Doing an ad-hoc array is way easier in that case and quite cheap: `[&space1, &space2]`.

`[&Space]` would satisfy the constrain above, but also forces downstream to create an ad-hoc array, which is annoying if you already have your references in another collection.
But on the other hand it is trivial to cast an array into a slice: `&[&space1, &space2]`.

That said, I think we can do better here.
@cmeissl Shouldn't `impl Iterator<Item=&Space>` also work here?

Reply:
Yeah, I think so, probably even `impl IntoIterator<Item=&Space>` should work. I will take a look.

Reply:
@bwidawsk I guess this is resolved now?